nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java

```diff
@@ -0,0 +1,322 @@
+            .build();
+
+    private static Supplier<Long> CURRENT_TIMESTAMP = System::currentTimeMillis;
+
+    private Serializer<String> stringSerializer = (v, o) -> o.write(v.getBytes(StandardCharsets.UTF_8));
+    private Serializer<Map<String, ListedEntity>> listedEntitiesSerializer
+            = (v, o) -> objectMapper.writeValue(new GZIPOutputStream(o), v);
+    private Deserializer<Map<String, ListedEntity>> listedEntitiesDeserializer
+            = v -> (v == null || v.length == 0) ? null
+            : objectMapper.readValue(new GZIPInputStream(new ByteArrayInputStream(v)), new TypeReference<Map<String, ListedEntity>>() {});
+
+    private final String componentId;
+    private final ComponentLog logger;
+    private final Scope scope;
+
+    /*
+     * The nodeId and mapCacheClient being used at the previous trackEntities method execution is captured,
+     * so that it can be used when resetListedEntities is called.
+     */
+    private String nodeId;
+    private DistributedMapCacheClient mapCacheClient;
+
+    ListedEntityTracker(String componentId, ComponentLog logger, Scope scope) {
+        this.componentId = componentId;
+        this.logger = logger;
+        this.scope = scope;
+    }
+
+    static void validateProperties(ValidationContext context, Collection<ValidationResult> results, Scope scope) {
+        final Consumer<PropertyDescriptor> validateRequiredProperty = property -> {
+            if (!context.getProperty(property).isSet()) {
+                final String displayName = property.getDisplayName();
+                results.add(new ValidationResult.Builder()
+                        .subject(displayName)
+                        .explanation(format("'%s' is required to use '%s' listing strategy", displayName, AbstractListProcessor.BY_ENTITIES.getDisplayName()))
+                        .valid(false)
+                        .build());
+            }
+        };
+        validateRequiredProperty.accept(ListedEntityTracker.TRACKING_STATE_CACHE);
+        validateRequiredProperty.accept(ListedEntityTracker.TRACKING_TIME_WINDOW);
+
+        if (Scope.LOCAL.equals(scope)) {
+            if (StringUtils.isEmpty(context.getProperty(NODE_IDENTIFIER).evaluateAttributeExpressions().getValue())) {
+                results.add(new ValidationResult.Builder()
+                        .subject(NODE_IDENTIFIER.getDisplayName())
+                        .explanation(format("'%s' is required to use local scope with '%s' listing strategy",
+                                NODE_IDENTIFIER.getDisplayName(), AbstractListProcessor.BY_ENTITIES.getDisplayName()))
+                        .build());
+            }
+        }
+    }
+
+    static void setCurrentTimestampSupplier(Supplier<Long> supplier) {
+        CURRENT_TIMESTAMP = supplier;
+    }
+
+    private String getCacheKey() {
+        switch (scope) {
+            case LOCAL:
+                return format("%s::%s::%s", getClass().getSimpleName(), componentId, nodeId);
```

Comment:
We should avoid using `getClass().getSimpleName()` here. If ever this class were to be renamed, it would reset all users' state for all instances. Instead, we should use a hardcoded value of "ListedEntityTracker".

Reply:
Done, I used 'ListedEntities' as prefix.